[dtensor] Fix and improve the sharding cache behavior

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109306

resolves https://github.com/pytorch/pytorch/issues/109101

The problem is essentially because we were hashing all the arguments, including
the scalar too (i.e. aten.div(tensor, scalar)), in the optimizer, the scalar might
change everytime we call the op, thus cache miss everytime we call the op

This PR improves the sharding cache behavior by introducing a
RuntimeSchemaInfo, used to record some runtime necessary hashing
information during op registration time. This enable us to:
* only hash arguments that are tensor or have static_argnum, this is to
enable many cases like aten.div.Tensor(tensor, 0.23231) hit the cache.
as we currently hashing all args which exclude those cases
* with the correct cache behavior, optimizers will hit the cache again
and resolve the high cpu overhead issue.

simple MLP shows all cache hit and for a single addmm -> 0.319ms (from 0.341ms), shows some hashing improvements:
<img width="1172" alt="Screenshot 2023-09-14 at 11 06 07 AM" src="https://github.com/pytorch/pytorch/assets/9443650/3406d673-dd8d-4ad9-9b80-9d4721c430e3">


Adam optimizer shows aten.div hit sharding cache again
<img width="1016" alt="Screenshot 2023-09-14 at 11 02 10 AM" src="https://github.com/pytorch/pytorch/assets/9443650/4280e8e3-af44-4fc2-8360-ea80b768f1d9">

wanchaol committed:
[dtensor] Fix and improve the sharding cache behavior

resolves https://github.com/pytorch/pytorch/issues/109101

This PR improves the sharding cache behavior by introducing a
RuntimeSchemaInfo, used to record some runtime necessary hashing
information during op registration time. This enable us to:
* only hash arguments that are tensor or have static_argnum, this is to
enable many cases like aten.div.Tensor(tensor, 0.23231) hit the cache.
as we currently hashing all args which exclude those cases
* with the correct cache behavior, optimizers will hit the cache again
and resolve the high cpu overhead issue.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] Fix and improve the sharding cache behavior"


resolves https://github.com/pytorch/pytorch/issues/109101

The problem is essentially because we were hashing all the arguments, including
the scalar too (i.e. aten.div(tensor, scalar)), in the optimizer, the scalar might
change everytime we call the op, thus cache miss everytime we call the op

This PR improves the sharding cache behavior by introducing a
RuntimeSchemaInfo, used to record some runtime necessary hashing
information during op registration time. This enable us to:
* only hash arguments that are tensor or have static_argnum, this is to
enable many cases like aten.div.Tensor(tensor, 0.23231) hit the cache.
as we currently hashing all args which exclude those cases
* with the correct cache behavior, optimizers will hit the cache again
and resolve the high cpu overhead issue.

simple MLP shows all cache hit and for a single addmm -> 0.319ms (from 0.341ms), shows some hashing improvements:
<img width="1172" alt="Screenshot 2023-09-14 at 11 06 07 AM" src="https://github.com/pytorch/pytorch/assets/9443650/3406d673-dd8d-4ad9-9b80-9d4721c430e3">


Adam optimizer shows aten.div hit sharding cache again
<img width="1016" alt="Screenshot 2023-09-14 at 11 02 10 AM" src="https://github.com/pytorch/pytorch/assets/9443650/4280e8e3-af44-4fc2-8360-ea80b768f1d9">



[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] Fix and improve the sharding cache behavior"


resolves https://github.com/pytorch/pytorch/issues/109101

The problem is essentially because we were hashing all the arguments, including
the scalar too (i.e. aten.div(tensor, scalar)), in the optimizer, the scalar might
change everytime we call the op, thus cache miss everytime we call the op

This PR improves the sharding cache behavior by introducing a
RuntimeSchemaInfo, used to record some runtime necessary hashing
information during op registration time. This enable us to:
* only hash arguments that are tensor or have static_argnum, this is to
enable many cases like aten.div.Tensor(tensor, 0.23231) hit the cache.
as we currently hashing all args which exclude those cases
* with the correct cache behavior, optimizers will hit the cache again
and resolve the high cpu overhead issue.

simple MLP shows all cache hit and for a single addmm -> 0.319ms (from 0.341ms), shows some hashing improvements:
<img width="1172" alt="Screenshot 2023-09-14 at 11 06 07 AM" src="https://github.com/pytorch/pytorch/assets/9443650/3406d673-dd8d-4ad9-9b80-9d4721c430e3">


Adam optimizer shows aten.div hit sharding cache again
<img width="1016" alt="Screenshot 2023-09-14 at 11 02 10 AM" src="https://github.com/pytorch/pytorch/assets/9443650/4280e8e3-af44-4fc2-8360-ea80b768f1d9">



[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -171,25 +190,80 @@ def __repr__(self) -> str:
+        if not is_list_like:
+            return False
+
+        elem_type = op_arg_type.getElementType()
+        return isinstance(elem_type, torch.TensorType) or (
+            isinstance(elem_type, torch.OptionalType)
+            and isinstance(elem_type.getElementType(), torch.TensorType)
+        )
```
For my learning purpose, what cases will trigger this logic?

when `isinstance(elem_type, torch.TensorType)`

wanchaol replied:
for argument types like `Tensor[]` or `Tensor?[]`, this is basically to test these routes and make sure we don't hash those arguments

fduwjj commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -130,6 +130,23 @@ def __str__(self) -> str:
+    # This static_argnum records static arg starting index for ops that have non-tensor
+    # args/kwargs which would affect sharding propagation results.
+    # only a few ops need this information, e.g. view, transpose, var.dim, etc.
+    static_argnum: int = -1
```
Does this we made the assumption that the first static_argnum args will affect sharding props, and args more than that is not? For a fix, this is fine as long as all DTensor unit test passes. But I am wondering if this might be a hacky assumption here.

wanchaol replied:
actually no, the comment says `static_argnum` is the starting index that would affect sharding prop, so the arguments on or after `static_argnum` would influence whether to run the sharding propagation again. I think this is the right way to handle caching uniformly and there's no violation so far in all the ops I skimmed

fduwjj reviewed (approved):
Is the `xla.py` expected? Overall, the idea looks make sense to me and accept to unblock for now. We might need to think more to make it less hacky down the road?

wanchaol commented:
> Is the `xla.py` expected? Overall, the idea looks make sense to me and accept to unblock for now. We might need to think more to make it less hacky down the road?

oops good catch, the `xla.py` is not expected.. I accidentally checked that in, will remove. I think this should be the right way to handle sharding cache in a uniform manner, the previous hashing method we had does not work with many corner cases like `aten.div.Scalar` and a bit bug prone. How custom op registration work with this is not flushed out and I agree we probably need some better way to handle custom ops.

wanchaol committed:
Update on "[dtensor] Fix and improve the sharding cache behavior"


resolves https://github.com/pytorch/pytorch/issues/109101

The problem is essentially because we were hashing all the arguments, including
the scalar too (i.e. aten.div(tensor, scalar)), in the optimizer, the scalar might
change everytime we call the op, thus cache miss everytime we call the op

This PR improves the sharding cache behavior by introducing a
RuntimeSchemaInfo, used to record some runtime necessary hashing
information during op registration time. This enable us to:
* only hash arguments that are tensor or have static_argnum, this is to
enable many cases like aten.div.Tensor(tensor, 0.23231) hit the cache.
as we currently hashing all args which exclude those cases
* with the correct cache behavior, optimizers will hit the cache again
and resolve the high cpu overhead issue.

simple MLP shows all cache hit and for a single addmm -> 0.319ms (from 0.341ms), shows some hashing improvements:
<img width="1172" alt="Screenshot 2023-09-14 at 11 06 07 AM" src="https://github.com/pytorch/pytorch/assets/9443650/3406d673-dd8d-4ad9-9b80-9d4721c430e3">


Adam optimizer shows aten.div hit sharding cache again
<img width="1016" alt="Screenshot 2023-09-14 at 11 02 10 AM" src="https://github.com/pytorch/pytorch/assets/9443650/4280e8e3-af44-4fc2-8360-ea80b768f1d9">



[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
